win32ss/gdi/ntgdi/freetype.c

```diff
@@ -2994,38 +2993,103 @@ static unsigned int get_bezier_glyph_outline(FT_Outline *outline, unsigned int b
 static INT
 IntRequestFontSize(PDC dc, FT_Face face, LONG Width, LONG Height)
 {
-    FT_Size_RequestRec  req;
+    FT_Error error;
+    ULONG TargetHeight, TargetWidth, RealHeight;
 
-    if (Width < 0)
-        Width = -Width;
+    ASSERT(face != NULL);
+    ASSERT(face->height > 0);
+    ASSERT(face->units_per_EM > 0);
+
+    if (Height == 0)
+    {
+        Height = dc->ppdev->devinfo.lfDefaultFont.lfHeight;
+    }
 
     if (Height < 0)
     {
-        Height = -Height;
+        /* case (A): lfHeight is negative: the character height */
+        Height = -Height << 6;
     }
-    if (Height == 0)
+    else if (Height > 0)
     {
-        Height = dc->ppdev->devinfo.lfDefaultFont.lfHeight;
+        /* case (B): lfHeight is positive: the cell height */
+        Height <<= 6;
+        Height = (Height * face->units_per_EM) / face->height;
     }
-    if (Height == 0)
+
+    TargetHeight = (Height * face->height) / face->units_per_EM;
+    TargetWidth = Width = Width << 6;
+
+    if (Width && FT_IS_SCALABLE(face))
     {
-        Height = Width;
+        if (face_has_encoding(face, FT_ENCODING_SJIS) ||
+            face_has_encoding(face, FT_ENCODING_PRC) ||
+            face_has_encoding(face, FT_ENCODING_BIG5) ||
+            face_has_encoding(face, FT_ENCODING_WANSUNG) ||
+            face_has_encoding(face, FT_ENCODING_JOHAB))
+        {
+            /* Case (i). Likely, the face has many fullwidth characters.
+               The width value was same as fullwidth character width. */
+        }
+        else
+        {
+            /* Case (ii). It's a Latin-like simple face.
+               The width value was same as halfwidth character width. */
+            Width *= 2;
+            TargetWidth *= 2;
```

Comment:
Or: `(Target)Width <<= 1;` :D

Reply:
Shush. Someone might take you seriously :p

Reply:
OK, I'll delete these comments and use <<= 2.

Reply:
The comments look fine I think.

Reply:
I did "s/*= 2/<<= 1/".

Reply:
Hermès was joking. Please don't use shifts where multiplication is the logical thing to do. The compiler knows that `*= 2` and `<<= 1` are the same thing and can optimize. So it's readability that counts.

Reply:
Reverted.